data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+import xml.etree.ElementTree as ET
+import requests
+import pandas as pd
+import datetime as dt
+import calendar
+from pathlib import Path
+import shutil
+import time
+import random
+
+from . import df_to_tempfile
+
+from typing import Literal, TypedDict, Any
+
+api_endpoint = "https://www.checkbooknyc.com/api"
+
+TYPE_OF_DATA = Literal[
+    "Contracts",
+    "Budget",
+    "Revenue",
+    "Payroll",
+    "Spending",
+    "Spending_OGE",
+    "Contracts_OGE",
+    "Spending_NYCHA",
+    "Contracts_NYCHA",
+    "Payroll_NYCHA",
+]
+
+DEFAULT_RECORDS_FROM = "1"
+DEFAULT_MAX_RECORDS = "20000"
+
+DEFAULT_DATE_FORMAT = "%Y-%m-%d"
+DEFAULT_START_PERIOD = dt.date(2010, 1, 1).strftime(DEFAULT_DATE_FORMAT)
+DEFAULT_END_PERIOD = dt.date.today().strftime(DEFAULT_DATE_FORMAT)
+
+DEFAULT_DATA_PATH = Path(__file__).resolve().parent / "capital_spending_data"
+
+USER_AGENTS = [
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+]
+user_agent = random.choice(USER_AGENTS)
+
+
+class CriteriaValue(TypedDict):
+    name: str
+    type: Literal["value"]
+    value: str
+
+
+class CriteriaRange(TypedDict):
+    name: str
+    type: Literal["range"]
+    start: str
+    end: str
+
+
+class Scriptor:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+    def ingest(
+        self,
+        user_agent: str | None = None,
+        type_of_data: TYPE_OF_DATA = "Spending",
+        records_from: str = DEFAULT_RECORDS_FROM,
+        max_records: str = DEFAULT_MAX_RECORDS,
+        response_columns: list[str] | None = None,
+        search_criteria: list[CriteriaValue | CriteriaRange] = [],
+        num_retries: int = 3,
+        data_dir: Path = DEFAULT_DATA_PATH,
+    ) -> pd.DataFrame:
+        """Fetches data from Checkbook NYC API based on input parameters, saves it
+        locally in parquet files organized by months, and returns data in a pandas DataFrame.
+        """
+        create_data_dir(data_dir)
+
+        start_period = DEFAULT_START_PERIOD
```

Comment:
Looking at the logic below, should these maybe just be supplied as their own args? Since they're a bit of a special case for search criteria. I like starting with search criteria as more generic - a user could supply whatever combination of search criteria they want, and we don't need to necessarily validate whatever combination they provide. But given that someone could arbitrarily provide multiple issue dates and we have special logic around ONE range of dates, I'd prefer to be explicit about that expectation.

For the checkbook api, are there other fields that support ranges? Or just issue date?

Reply:
@fvankrieken, makes sense! I will move start and end issue dates to its own args. 

Do we still want to validate that none were provided in the search criteria? I could see having conflicting issue dates if provided in 2 places...

Regarding the ranges... Yes, the other field that supports it in the `Spending` domain is `check_amount `. Note, "range" fields may vary across the six domains because columns across them are not uniform.